File: Leetcode_Code/113. Path Sum II/113.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        if (root == null) return result;
        Stack<Integer> stack = new Stack<>();
        pathSum_dfs(root, stack, sum);
        return result;
    }

    private List<List<Integer>> result = new ArrayList<>();

    private void pathSum_dfs(TreeNode root, Stack<Integer> stack, int sum) {
        if (root.left == null & root.right == null) {
            if (sum == root.val) {
                stack.add(root.val);
                List<Integer> object = new ArrayList<>(stack);
                result.add(object);
                stack.pop();
            }
        }
        stack.add(root.val);
        if (root.left != null) pathSum_dfs(root.left, stack, sum - root.val);
        if (root.right != null) pathSum_dfs(root.right, stack, sum - root.val);
        stack.pop();
    }
}

File: Leetcode_Code/113. Path Sum II/113.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.result = []

    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if root:
            self.pathSum_dfs(root, [], sum)
        return self.result

    def pathSum_dfs(self, root, stack, sum):
        if root.left is None and root.right is None:
            if sum == root.val:
                self.result.append(stack + [root.val])
            return
        stack.append(root.val)
        if root.left is not None:
            self.pathSum_dfs(root.left, stack, sum - root.val)
        if root.right is not None:
            self.pathSum_dfs(root.right, stack, sum - root.val)
        stack.pop()
